create reusable modal alert

kimberlythegeek opened the pull request:
Added dismissable alert to Header component. This replaces the `console.log` statement when a user tries to select more than 4 revisions. Currently, because of the way the focus hook works, dismissing the notification will also hide the search results. I'd like to replace this with a timeout, or at least not hide the results somehow

kimberlythegeek commented:
blocked by #102

eng-esther commented on src/hooks/useCheckRevision.ts:
```diff
@@ -0,0 +1,31 @@
+    } else if (currentIndex !== -1) {
+      newChecked.splice(currentIndex, 1);
+    } else {
+      useSetAlert('Maximum 4 Revisions', 'warning');
```
is this alert timed? or it will remain on the page until it has been addressed?

eng-esther commented on src/hooks/useCheckRevision.ts:
```diff
@@ -0,0 +1,31 @@
+    } else if (currentIndex !== -1) {
+      newChecked.splice(currentIndex, 1);
+    } else {
+      useSetAlert('Maximum 4 Revisions', 'warning');
```
is this alert timed? or it will remain on the page until it has been addressed?

kimberlythegeek replied:
as it is now, it will stay until dismissed.

kimberlythegeek commented:
> After the alert appears I can't uncheck the checked revisions from the list.

ah--thanks for catching that. I modified the if/else statement and apparently did it wrong :) should be fixed now. Also added a test case to catch this.

julienw commented on src/hooks/useAlert.ts:
```diff
@@ -0,0 +1,25 @@
+    dispatch(setAlert({ message, title, severity }));
+  };
+
+  return { useClearAlert, useSetAlert };
```
I find the names a bit confusing: they look like they are hooks themselves, but they actually dispatch an action.
I would use simply `clearAlert` and `setAlert`, or more explicit `dispatchClearAlert` and `dispatchSetAlert`. (but I prefer the first option).

kimberlythegeek replied:
I can't use `clearAlert` and `setAlert` as those are the names of the reducer actions, but I can change it to `dispatch` instead of `set`

julienw commented on src/hooks/useAlert.ts:
```diff
@@ -0,0 +1,25 @@
+import { clearAlert, setAlert } from '../reducers/AlertSlice';
+import type { AlertType } from '../types/state';
+
+const useAlert = () => {
```
optional comment: is this hook really useful? I feel like it's adding more code than it removes.

kimberlythegeek replied:
I guess I don't have a strong argument for it now, but ideally in the future we will be able to display more than one alert at a time, and this hook would be used to handle those actions.

julienw commented on src/hooks/useCheckRevision.ts:
```diff
@@ -11,18 +13,17 @@ const useCheckRevision = () => {
     const currentIndex = checked.indexOf(revisionID);
```
optional: It would be clearer to have a boolean if you dont use the index itself. You can use `const isAlreadyChecked = checked.includes(revisionID)` for example.

julienw commented on src/hooks/useCheckRevision.ts:
```diff
@@ -1,8 +1,10 @@
 const useCheckRevision = () => {
+  const { useSetAlert } = useAlert();
   const { handleChildClick } = useFocusInput();
   const [checked, setChecked] = React.useState<string[]>([]);
```
optional: I wonder if using a `Set<string>` could be a better fit, seeing the operations you're doing here.

kimberlythegeek replied:
I'll leave this as is for now, and file an issue to consider using Sets instead. it seems like it's worthwhile when handling many items, but this array will never have more than 4 elements. at least for while :) #137

julienw replied:
Yeah, to be clear, this wasn't a performance-oriented reason, more a simplification of the code, where you wouldn't need to check for existence before pushing :-)

julienw commented on src/tests/PerfCompareHeader.test.tsx:
```diff
@@ -0,0 +1,44 @@
+    render(<PerfCompareHeader />);
+
+    expect(screen.getByText('Maximum 4 Revisions')).toBeInTheDocument();
+    expect(screen).toMatchSnapshot();
```
Usually I use `document.body` to match.
Looking at the new snapshot, I think that using `screen` doesn't work as you expect :-)

julienw commented on src/components/Shared/HeaderAlert.tsx:
```diff
@@ -0,0 +1,34 @@
+        useClearAlert();
+      }}
+    >
+      {alert.title && <AlertTitle>{alert.title}</AlertTitle>}
+      {alert.message}
```
What happens if you have `isAlert === true` but both title and message are empty? Should we display some default message?

kimberlythegeek replied:
`title` is the only parameter with a default value, so if you try to call `useSetAlert` without a message, it will throw an error.

I could add a default message though, as there may be a use case in the future where an unexpected error occurs.

julienw commented on src/components/Shared/HeaderAlert.tsx:
```diff
@@ -0,0 +1,34 @@
+  return (
+    <Alert
+      severity={alert.severity}
+      onClose={() => {
```
It would probably be a good idea to call `e.stopPropagation()` on the event passed to the callback, so that the close operation doesn't trigger other things (eg: so that it doesn't close the select box)

kimberlythegeek replied:
thanks!

julienw reviewed (approved):
Thanks, this looks good to me!
I left a few comments, some of them are optional too, so tell me what you think!
